subsys/bluetooth/shell/mics_client.c

```diff
@@ -0,0 +1,625 @@
+			    "inst %p", err, inst);
+	} else {
+		shell_print(ctx_shell, "AICS inst %p description %s",
+			    inst, description);
+	}
+}
+
+static struct bt_mics_cb mics_cbs = {
+	.discover = mics_discover_cb,
+	.mute_write = mics_mute_write_cb,
+	.mute = mics_mute_cb,
+
+	/* Audio Input Control Service */
+	.aics_cb = {
+		.state = mics_aics_state_cb,
+		.gain_setting = mics_aics_gain_setting_cb,
+		.type = mics_aics_input_type_cb,
+		.status = mics_aics_status_cb,
+		.description = mics_aics_description_cb,
+		.set_gain = mics_aics_set_gain_cb,
+		.unmute = mics_aics_unmute_cb,
+		.mute = mics_aics_mute_cb,
+		.set_manual_mode = mics_aics_set_manual_mode_cb,
+		.set_auto_mode = mics_aics_automatic_mode_cb,
+	}
+};
+
+static int cmd_mics_client_discover(const struct shell *sh, size_t argc,
+				    char **argv)
+{
+	int result;
+
+	if (ctx_shell == NULL) {
+		ctx_shell = sh;
+	}
+
+	result = bt_mics_client_cb_register(&mics_cbs);
+	if (result != 0) {
+		shell_print(sh, "Failed to register callbacks: %d", result);
+	}
+
+	if (default_conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	result = bt_mics_discover(default_conn);
+	if (result != 0) {
+		shell_print(sh, "Fail: %d", result);
+	}
+
+	return result;
+}
+
+static int cmd_mics_client_mute_get(const struct shell *sh, size_t argc,
+				    char **argv)
+{
+	int result;
+
+	if (default_conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	result = bt_mics_mute_get(default_conn);
+
+	if (result != 0) {
+		shell_print(sh, "Fail: %d", result);
+	}
+
+	return result;
+}
+
+static int cmd_mics_client_mute(const struct shell *sh, size_t argc,
+				char **argv)
+{
+	int result;
+
+	if (default_conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	result = bt_mics_mute(default_conn);
+
+	if (result != 0) {
+		shell_print(sh, "Fail: %d", result);
+	}
+
+	return result;
+}
+
+static int cmd_mics_client_unmute(const struct shell *sh, size_t argc,
+				  char **argv)
+{
+	int result;
+
+	if (default_conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	result = bt_mics_unmute(default_conn);
+
+	if (result != 0) {
+		shell_print(sh, "Fail: %d", result);
+	}
+
+	return result;
+}
+
+static int cmd_mics_client_aics_input_state_get(const struct shell *sh,
+						size_t argc, char **argv)
+{
+	int result;
+	int index = strtol(argv[1], NULL, 0);
+
+	if (index >= mics.aics_cnt) {
+		shell_error(sh, "Index shall be less than %u, was %u",
+			    mics.aics_cnt, index);
+		return -ENOEXEC;
+	}
+
+	if (default_conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	result = bt_mics_aics_state_get(default_conn, mics.aics[index]);
+	if (result != 0) {
+		shell_print(sh, "Fail: %d", result);
+	}
+
+	return result;
+}
+
+static int cmd_mics_client_aics_gain_setting_get(const struct shell *sh,
+						 size_t argc, char **argv)
+{
+	int result;
+	int index = strtol(argv[1], NULL, 0);
+
+	if (index >= mics.aics_cnt) {
+		shell_error(sh, "Index shall be less than %u, was %u",
+			    mics.aics_cnt, index);
+		return -ENOEXEC;
+	}
```

Comment:
The work is already done now - but I think there could be some typing (or copying and pasting) and some code lines saved by putting the index creation and check into a macro or two in these files.

Reply:
I'd rather not have flow control (i.e. return statements) in macros; we could add a function that checks the conn pointer and index value, and check the return value to write the proper error message to the shell; I don't think we'd actually gain anything from that (in terms of fewer lines), except moving some common functionality (which arguably is of course a good thing).

Reply:
I agree - flow control in macros can be confusing.
I was thinking it could be solved by giving the macro a very explicit name, like RETURN_IF_INVALID_INDEX or similar.
(Or possibly, with the side effect being only the output:
```
if (INVALID_INDEX) {
        return -ENOEXEC;
}
```
But I am OK leaving this as is.